### PR TITLE
Implemented CanCanCan

### PR DESCRIPTION
**Pull request by VelzckC0D3**
# ✨ Requested Implementations ✨

### ✔️ _Installed and Prepared all the CanCanCan environment_
- _Added role column to the User table using migration_
- _A user can delete its own posts, and the admin user can delete any post_
_For this feature, I've implemented the deletion button with its own route, which checks the user role_
- _A user can delete its own comments, and the admin user can delete any comment_
_For this feature, I've implemented the deletion button with its own route, this checks the user role_

### 🎈 _Testing video_
- _In this short video you can check how the authentication system works_

https://github.com/VelzckC0D3/blog_app_rails/assets/92229666/e3ae7a4b-3329-48e2-99ab-3935746a4518

# 💡 Thank you for your review! 💡

### CODE REVIEW

File: app/models/ability.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    can :create, :user
+
+    can :manage, :all if user.role == 'admin'
+
+    return unless user.role == 'user'
+
+    can :create, Post
+    can :create, Comment
+    can :destroy, Comment
+    can :destroy, Post
```
Comment (anagudelogu) [inline app/models/ability.rb]: - [ ] Please ensure that only authors of Posts and Comments can destroy them. At the moment, you're only hiding the UI buttons for these actions. I suggest something like this: 🤓 🔧 

```rb
# owner of post can perform all view, update or destroy own post if Post.author == user
         can :manage, Post, author: user 
# owner of comment can perform all view, update or destroy own comment if Comment.author == user
         can :manage, Comment, author: user
```

File: app/views/posts/show.html.erb [hunk 2]
```diff
@@ -1,14 +1,20 @@
...
   <div class="posts-show-actions">
-    <%= button_to 'Go Back', user_posts_path(@post.user), method: :get, class: 'posts-show-btn' %>
+    <%= button_to 'Go Back', user_posts_path(@post.user), method: :get, class: 'posts-show-btn', data: { turbo: false } %>
     <%= button_to 'Go Home', users_path, method: :get, class: 'posts-show-btn' %>
-    <%= button_to 'New Comment', new_comment_path(@post.user), method: :get, class: 'posts-show-btn' %>
+    <% if user_signed_in? %>
+      <% if can? :create, Comment  %>
+        <%= button_to 'New Comment', new_comment_path(@post.user), method: :get, class: 'posts-show-btn' %>
+      <% end %>
+      <% if can? :destroy, Post && current_user.id == @post.user.id or current_user.id == @post.user.id   %>
```
Comment (anagudelogu) [inline app/views/posts/show.html.erb]: - [ ] **[OPTIONAL]** When you implement the previous change, this condition will be redundant, so only calling the `can?` method will be enough to check for permissions:

By checking
```rb
if can? :destroy, Post
```

It calls the Ability class initialize method and checks if the `Post.user.id == current_user.id`. Please consider applying this change! 🤓 🔧

**Review by anagudelogu — CHANGES_REQUESTED** [review]
# STATUS:  CHANGES REQUIRED  ♻️ 🚧 

Hi @VelzckC0D3 @Danfall369, 
I'm Andres.

While you have done a great job working on this milestone 💻 , there are some issues that you still need to work on to go to the next project. ♻️ 

You are almost there, keep up the good work! 🚀 



## Highlights
✅ Great use of Git flow. 💯 
✅ Your project is documented professionally. 📜 
✅ Great UI design. 💻 
✅ Descriptive PR title, summary and commit messages. 💬 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/views/posts/show.html.erb (hunk 2 above)
Comment (VelzckC0D3) [reply app/views/posts/show.html.erb]: @anagudelogu Thank you for letting me know, I've also corrected that!

File: app/models/ability.rb (hunk 1 above)
Comment (VelzckC0D3) [reply app/models/ability.rb]: @anagudelogu Got it! thank you for noticing, I've already made the changes

**Review by Owusu-Desmond — APPROVED** [review]
Hi @VelzckC0D3 , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
📌 No Linters error ✅
📌 Descriptive pull request✅
📌 Good commit messages ✅
📌 Professional README file ✅

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**